### PR TITLE
Deblur and Denoise using Richardson Lucy Deconvolution for Computer Vision tasks

### PR DESCRIPTION
**Pull request by rzimmerdev**
### Describe your change:
I've added the Richardson Lucy deconvolution algorithm, as well as some supporting auxiliary methods.
This algorithm is useful for recovering degraded images, given a known noise and/or blurring function.

* [x] Add an algorithm?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [X] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by alleycat786 — APPROVED** [review]
tq so much!

**Review by alleycat786 — APPROVED** [review]
tq so much!

**Review by alleycat786 — APPROVED** [review]
tq so muchh! good work

**Review by alleycat786 — APPROVED** [review]
tq!

**Comment by alleycat786** [comment]
tq so much good work

**Review by alleycat786 — APPROVED** [review]
tq so much

File: computer_vision/richardson_lucy.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        >>> degraded = gaussian_noise(shape)
+        >>> kernel = np.identity(2)
+        >>> richardson_lucy(degraded, kernel, steps=4)
+        array([[1.16272775e+02, 0.00000000e+00, 7.58091075e+01],
+               [2.40582123e+01, 2.09006034e+01, 0.00000000e+00],
+               [0.00000000e+00, 1.52620152e-02, 1.31734948e+00]])
+    """
+    estimated_img = np.full(shape=degraded.shape, fill_value=1, dtype="float64")
+
+    for i in range(steps):
```
Comment (tianyizheng02) [inline computer_vision/richardson_lucy.py:213]:
```suggestion
    for _ in range(steps):
```
`i` is never used in the loop body, and this is causing the pre-commit to fail

File: computer_vision/haralick_descriptors.py
Comment (tianyizheng02) [inline computer_vision/haralick_descriptors.py:1]: Wrong PR, already included in #8004

File: computer_vision/richardson_lucy.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    arx = np.arange((-k // 2) + 1.0, (k // 2) + 1.0)
+    x, y = np.meshgrid(arx, arx)
+    filt = np.exp(-(1 / 2) * (np.square(x) + np.square(y)) / np.square(sigma))
+    return filt / np.sum(filt)
+
+
+def convolve(matrix: np.ndarray, kernel: np.ndarray) -> np.ndarray:
+    """
+    Convolves a given kernel around a matrix through the frequency domain,
+    using Fourier transformations.
```
Comment (tianyizheng02) [inline computer_vision/richardson_lucy.py:120]:
```suggestion
    using Fourier transforms.
```
Nitpick: "Fourier transforms", not "Fourier transformations"

File: computer_vision/richardson_lucy.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    Examples:
+        >>> matrix = np.array([[-1.45436567,  0.04575852],
+        ...                    [-0.18718385,  1.53277921]])
+        >>> kernel = np.array([[1, 0], [0, 1]])
+        >>> convolve(matrix, kernel)
+        array([[ 0.07841354, -0.14142533],
+               [-0.14142533,  0.07841354]])
+    """
+    if kernel.shape[0] > matrix.shape[1] or kernel.shape[1] > matrix.shape[1]:
+        return matrix
```
Comment (tianyizheng02) [inline computer_vision/richardson_lucy.py:139]: If the kernel is too big, why not raise an error?

File: computer_vision/richardson_lucy.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+    return estimated_img
+
+
+def main():
+    input_file = str(input()).rstrip()
+    degraded = imageio.imread(input_file)
+
+    angle = int(input())
+    steps = int(input())
```
Comment (tianyizheng02) [inline computer_vision/richardson_lucy.py:231]: Can we get prompt messages for the `input` calls

File: computer_vision/richardson_lucy.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        0.816496580927726
+        >>> root_mean_square_error(np.array([1, 2, 3]), np.array([6, 4, 2]))
+        3.1622776601683795
+    """
+    return np.sqrt(((original - reference) ** 2).mean())
+
+
+def pad_to_size(image: np.ndarray, reference: np.ndarray):
+    """Pad an image to have final shape equal to reference image."""
+    p, q = (size for size in reference.shape)
```
Comment (tianyizheng02) [inline computer_vision/richardson_lucy.py:28]: Are these variables necessary? Why not just use `reference.shape[0]` and `reference.shape[1]`?

File: computer_vision/richardson_lucy.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        return matrix
+    kernel = pad_to_size(kernel, matrix)
+
+    kernel_f = fft2(kernel)
+    matrix_f = fft2(matrix)
+
+    return ifftshift(ifft2(np.multiply(matrix_f, kernel_f))).real
+
+
+def get_motion_psf(shape: tuple, angle: float, num_pixel_dist: int = 20) -> np.ndarray:
```
Comment (tianyizheng02) [inline computer_vision/richardson_lucy.py:148]:
```suggestion
def get_motion_psf(shape: tuple[int, int], angle: float, num_pixel_dist: int = 20) -> np.ndarray:
```

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

**Comment by tianyizheng02** [comment]
This PR is unfortunately taking too long and too much effort to fix, and I don't have the time because we maintainers have to clear out the backlog of PRs as much as possible before Hacktoberfest. In addition, it appears that this PR has failing doctests even when the code formatting and styling has been fixed.

**Pull request closed without merging**